### PR TITLE
[for testing] Add minimum mesh generation interval

### PR DESCRIPTION
**Pull request by numberZero**
See #3602

### CODE REVIEW

File: builtin/settingtypes.txt [hunk 1]
```diff
@@ -508,6 +508,8 @@ hud_hotbar_max_width (Maximum hotbar width) float 1.0
 #    Enables caching of facedir rotated meshes.
 enable_mesh_cache (Mesh cache) bool false
 
+mesh_generation_interval (Mesh generation interval) int 15 0 50
```
Comment (sofar) [inline builtin/settingtypes.txt]: No units? Is higher better or worse? How did you choose the default?

Comment (paramat) [inline builtin/settingtypes.txt]: This is only for testing at the moment so doesn't need to be merge-worthy.

**Comment by Fixer-007** [comment]
Units in milliseconds (ms), smaller value - bigger jitten, bigger value - smaller jitter

**Comment by lhofhansl** [comment]
This severely slows down map updates on the clients. The default should be 0.

**Comment by lhofhansl** [comment]
Otherwise (with default of 0) :+1:

**Comment by Fixer-007** [comment]
I'm fine with default at 0, I would like to see this merged. People affected with jitter will simply increase the sleep number to their liking.

File: builtin/settingtypes.txt (hunk 1 above)
Comment (numberZero) [inline builtin/settingtypes.txt]: @Fixer-007 can you write a description for this? I’m not very good in documenting.

**Comment by lhofhansl** [comment]
Documentation something perhaps like
"Delay between mesh updates on the client in ms. Increasing this will slow the rate of mesh updates, which will reduce jitter on slower clients"?

Also after some more reflection, should we default this to 10ms perhaps? I see the real slowdown only with larger v_range and when flying in fast mode.

**Comment by numberZero** [comment]
@lhofhansl, thank you.
@sofar, @Fixer-007 seems to want this to be merged.

**Comment by paramat** [comment]
But if merged this would be a possibly temporary workaround? Sounds like some people want to look for a new solution.

**Comment by numberZero** [comment]
> temporary workaround

@paramat sure.

> some people want to look for a new solution.

I do =) Sadly I haven’t enough time for that right now, the code is multi-threaded thus complicated... but I have some plans already anyway)

**Comment by lhofhansl** [comment]
The "real" fix is to allow true asynchronous rendering. I might be wrong, but that might be akin to a rewrite of Minetest.

Perhaps a starting point is to make MapBlocks immutable on the client, then at least we can take a set of MapBlocks and be sure they won't be modified after the fact.

I think this can be merged as a workaround.

**Comment by lhofhansl** [comment]
This should be added to minetest.conf.example as well.

**Comment by Fixer-007** [comment]
Tested this in "just test" server map save with tons of stuck liquids (that unstuck the moment I approach) and spawn chests that lagged insanely, with sleep at 15ms it makes a huuuge difference in playability, very smooth. Graph can look like this:
default

**Comment by lhofhansl** [comment]
Just be pedantic: This does _strictly_ process 1 mesh per sleep. It sleeps every time it pops a mesh from the queue.
With 15ms of sleep that still means up 66 meshes can be processed per second, that's what you're seeing.

@Fixer-007 Does the jigger in any way get better or worse depending on RTT to server (i.e. server lag). I wonder whether that has to do with ack'ing blocks incorrectly on the wrong thread.

**Comment by Fixer-007** [comment]
@lhofhansl 
IMO number of incoming packets does correlate with jitter, If I have few packets in - it plays pretty smooth, if no packets at all for some time - literally no jitter, if lots of packets in - lots of jitter.

**Comment by numberZero** [comment]
@lhofhansl 
> This should be added to minetest.conf.example as well.

done

**Comment by lhofhansl** [comment]
:+1: looks good, let's get it in.

**Comment by Fixer-007** [comment]
Nothing to see here https://i.imgur.com/arKqaoS.png

**Comment by sofar** [comment]
https://www.youtube.com/watch?v=euoDR6Iq73o

(May take a second to show up)

Seems to work well. Zooming around at high speed having terrain generated, I didn't notice any issues at all.

**Comment by sofar** [comment]
Note: I played with `0` - the default (no sleep).

**Comment by sofar** [comment]
Played with values like `5` and up to `50`, and this does exactly what it does. I even feel like `5` actually would be a reasonable default, and it reminds me a lot of how MC pulls in blocks.

:+1:

**Comment by paramat** [comment]
4d5177ff708c7e696eead18200e240047ff520fe

**Pull request closed without merging**